File: ehk-model/model/event.go
package model

import (
	"os"
	"sync"

	"github.com/vmihailenco/msgpack/v5"
)

// EventLogger handles logging of events during simulation
type EventLogger struct {
	Filename  string
	BatchSize int
	queue     chan any
	stopFlag  chan struct{}
	wg        sync.WaitGroup
	lock      sync.Mutex
	file      *os.File
}

// NewEventLogger creates a new event logger
func NewEventLogger(filename string, batchSize int) (*EventLogger, error) {
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		return nil, err
	}

	logger := &EventLogger{
		Filename:  filename,
		BatchSize: batchSize,
		queue:     make(chan any, batchSize*2),
		stopFlag:  make(chan struct{}),
		file:      file,
	}

	logger.wg.Add(1)
	go logger.worker()

	return logger, nil
}

// LogEvent adds an event to the queue for logging
func (l *EventLogger) LogEvent(event any) {
	select {
	case l.queue <- event:
		// Event queued successfully
	case <-l.stopFlag:
		// Logger is stopping, discard event
	}
}

// worker processes the event queue
func (l *EventLogger) worker() {
	defer l.wg.Done()

	batch := make([]any, 0, l.BatchSize)

	for {
		select {
		case event := <-l.queue:
			batch = append(batch, event)

			if len(batch) >= l.BatchSize {
				l.writeBatch(batch)
				batch = make([]any, 0, l.BatchSize)
			}

		case <-l.stopFlag:
			// Write any remaining events
			if len(batch) > 0 {
				l.writeBatch(batch)
			}
			return
		}
	}
}

// writeBatch writes a batch of events to the file
func (l *EventLogger) writeBatch(batch []any) {
	l.lock.Lock()
	defer l.lock.Unlock()

	for _, event := range batch {
		data, err := msgpack.Marshal(event)
		if err != nil {
			// Log error or handle it as appropriate
			continue
		}

		_, err = l.file.Write(data)
		if err != nil {
			// Log error or handle it as appropriate
		}
	}

	l.file.Sync()
}

// Stop closes the logger
func (l *EventLogger) Stop() {
	l.lock.Lock()

	// Signal worker to stop
	close(l.stopFlag)

	// Wait for worker to finish
	l.lock.Unlock()
	l.wg.Wait()

	// Close the file
	l.lock.Lock()
	defer l.lock.Unlock()
	l.file.Close()
}

// ReadMsgpackObjects reads msgpack objects from a file
func ReadMsgpackObjects(filename string) ([]any, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var objects []any
	decoder := msgpack.NewDecoder(file)

	for {
		var obj any
		err := decoder.Decode(&obj)
		if err != nil {
			break
		}
		objects = append(objects, obj)
	}

	return objects, nil
}

File: ehk_model_old/base/event.py
from typing import List, Generic, TypeVar

from io import BufferedReader, BufferedWriter
import queue
import threading
import msgpack
import atexit

T = TypeVar('T')


class EventLogger(Generic[T]):

  def __init__(self, filename: str, batch_size=1000):
    self.filename = filename
    self.batch_size = batch_size

    self.queue = queue.Queue()
    self.stop_flag = threading.Event()
    self.worker_thread = threading.Thread(target=self._worker)
    self.worker_thread.daemon = True
    self.worker_thread.start()
    self.lock = threading.Lock()

    atexit.register(self.stop)

  def log_event(self, event: T):
    self.queue.put(event)

  def _worker(self):
    with open(self.filename, 'ab') as f:
      batch = []
      while not self.stop_flag.is_set() or not self.queue.empty():
        try:
          event = self.queue.get(timeout=1)
          batch.append(event)
          if len(batch) >= self.batch_size:
            self._write_batch(f, batch)
            batch = []
        except queue.Empty:
          if batch:
            self._write_batch(f, batch)
            batch = []

  def _write_batch(self, file: BufferedWriter, batch: List[T]):
    with self.lock:
      for event in batch:
        msgpack.dump(event, file)
      file.flush()

  def stop(self):
    with self.lock:
      if not self.stop_flag.is_set():
        self.stop_flag.set()
        self.worker_thread.join(timeout=5)

        # force write any remaining events
        if not self.queue.empty():
          with open(self.filename, 'ab') as f:
            while not self.queue.empty():
              batch = []
              while not self.queue.empty() and len(batch) < self.batch_size:
                batch.append(self.queue.get_nowait())
              self._write_batch(f, batch)


def read_msgpack_objects(file: BufferedReader):
  unpacker = msgpack.Unpacker(file, raw=False)
  for obj in unpacker:
    yield obj
